src/Numerics/DGmethods/DGmodel_kernels.jl

```diff
@@ -79,162 +77,172 @@ See [`odefun!`](@ref) for usage.
         l_Qhypervisc = MArray{Tuple{nhyperviscstate}, FT}(undef)
         l_aux = MArray{Tuple{nauxstate}, FT}(undef)
         l_F = MArray{Tuple{3, nstate}, FT}(undef)
+        l_F3 = MArray{Tuple{nstate}, FT}(undef)
     end
 
-    s_F = @localmem FT (3, Nq, Nq, Nqk, nstate)
-    s_ω = @localmem FT (Nq,)
+    s_F = @localmem FT (2, Nq, Nq, nstate)
     s_D = @localmem FT (Nq, Nq)
-    l_rhs = @private FT (nstate,)
+
+    l_rhs = @private FT (Nqk, nstate)
+    l_MI = @private FT (Nqk,)
 
     e = @index(Group, Linear)
-    ijk = @index(Local, Linear)
-    i, j, k = @index(Local, NTuple)
+    i, j = @index(Local, NTuple)
 
     @inbounds begin
-        s_ω[j] = ω[j]
         s_D[i, j] = D[i, j]
-
-        M = vgeo[ijk, _M, e]
-        ξ1x1 = vgeo[ijk, _ξ1x1, e]
-        ξ1x2 = vgeo[ijk, _ξ1x2, e]
-        ξ1x3 = vgeo[ijk, _ξ1x3, e]
-        if dim == 3 || (dim == 2 && direction isa EveryDirection)
-            ξ2x1 = vgeo[ijk, _ξ2x1, e]
-            ξ2x2 = vgeo[ijk, _ξ2x2, e]
-            ξ2x3 = vgeo[ijk, _ξ2x3, e]
-        end
-        if dim == 3 && direction isa EveryDirection
-            ξ3x1 = vgeo[ijk, _ξ3x1, e]
-            ξ3x2 = vgeo[ijk, _ξ3x2, e]
-            ξ3x3 = vgeo[ijk, _ξ3x3, e]
+        @unroll for k in 1:Nqk
+            ijk = i + Nq * ((j - 1) + Nq * (k - 1))
+            @unroll for s in 1:nstate
+                l_rhs[k, s] = increment ? rhs[ijk, s, e] : zero(FT)
+            end
+            l_MI[k] = vgeo[ijk, _MI, e]
         end
 
-        @unroll for s in 1:nstate
-            l_rhs[s] = increment ? rhs[ijk, s, e] : zero(FT)
-        end
+        @unroll for k in 1:Nqk
+            @synchronize
```

Comment:
Is there any advantage to putting the `@synchronize` here versus right before the next set of shared writes?

Reply:
(I'm never sure the optimal way to do these since you need all the threads to be in this section, but they do not really have to wait until later)

Reply:
I guess until we are above order 5 it won't matter anyway, since we just have a single warp in the threadblock, so all the threads are in lock step anyway.

Reply:
`@synchronize` is there mainly because on the CPU I want to start the `ij` loop nest at the beginning of the `k` loop. However, I checked, and as you say, putting it further down doesn't seem to have an effect on the GPU for polynomial order 4.